Copyable as a Requirement Against the Machine

kavon opened the pull request:
This PR introduces the `NoncopyableGenerics` feature flag and should be a non-functional change when that feature is not enabled.

When the flag is enabled:
- Type parameters gain an implicit `Requirement` to be `Copyable`, rather than synthesizing extra constraints in the solver.
- Rudimentary support for removing that `Copyable` requirement via `~Copyable` from a protocol, associatedtype, or generic type parameter.
- Using `~Copyable` in the inheritance clause of a struct/enum now doesn't actually make the type recognized as noncopyable to the rest of the compiler. That's because we need to move away from the MoveOnlyAttr. This will be fixed in another PR.

This changeset is probably best reviewed one-commit-at-a-time.

resolves rdar://110420673&115812856

kavon committed:
`_Copyable` as a Requirement Against the Machine

An initial implementation of a rework in how
we prevent noncopyable types from being
substituted in places they are not permitted.

Instead of generating a constraint for every
generic parameter in the solver, we produce
real Copyable conformance requirements. This
is much better for our longer-term goal of
supporting `~Copyable` in more places.

slavapestov commented on lib/AST/RequirementMachine/RequirementLowering.cpp:
```diff
@@ -759,6 +759,20 @@ void swift::rewriting::realizeInheritedRequirements(
+    // Add Copyable as an "inferred" requirement.
+    SmallVector<Requirement, 2> reqs;
+    Requirement sugaredReq(RequirementKind::Conformance, type, copyableType);
+    desugarRequirement(sugaredReq, loc, reqs, errors);
```
desugarRequirement() is going to be a no-op in this case. you can just add the new requirement to `result`

kavon replied:
thank you!

slavapestov commented:
Also `didDiagnoseMoveOnlyGenericArgs()` in TypeCheckType.cpp is now redundant right?

slavapestov commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1817,6 +1817,11 @@ static void checkProtocolRefinementRequirements(ProtocolDecl *proto) {
 
+    // For any protocol 'P', there is an implied requirement 'Self: Copyable',
+    // unless it was suppressed via `Self: ~Copyable`; so skip if present.
+    if (otherProto->isSpecificProtocol(KnownProtocolKind::Copyable))
```
(And if it was suppressed with ~Copyable, falling through this code path will detect the case where something else implies it)

kavon committed:
[Sema] reimplement `~C` as an general inverse constraint

kavon committed:
[stdilb] introduce `Copyable` protocol

This replaces the old, unavailable `_Copyable`.

rdar://110420673

kavon committed:
[Generics] `~Copyable` in inheritance clauses

Have `~Copyable` change the signatures of a generic type param,
protocol, or associated type if written in the inheritance clause
position.

kavon committed:
[nfc] rename or eliminate `isPureMoveOnly` APIs

I think from SIL's perspective, it should only worry about whether the
type is move-only. That includes MoveOnlyWrapped SILTypes and regular
types that cannot be copied.

Most of the code querying `SILType::isPureMoveOnly` is in SILGen, where
it's very likely that the original AST type is sitting around already.
In such cases, I think it's fine to ask the AST type if it is
noncopyable. The clarity of only asking the ASTType if it's noncopyable
is beneficial, I think.

kavon commented on lib/AST/RequirementMachine/RequirementLowering.cpp:
```diff
@@ -756,6 +756,22 @@ void swift::rewriting::realizeInheritedRequirements(
+  auto &ctx = dc->getASTContext();
+  if (ctx.LangOpts.hasFeature(Feature::NoncopyableGenerics)) {
+    // All associated types or generic type params are Copyable by default.
+    // TODO: if we saw ~Copyable in the inherited types list earlier, skip this!
```
I think handling `where` is going to happen in a separate PR.

kavon commented:
@swift-ci please smoke test

kavon commented:
@swift-ci please smoke test

kavon commented:
@swift-ci please smoke test linux platform

slavapestov commented on include/swift/AST/KnownProtocols.h:
```diff
@@ -41,6 +42,12 @@ enum : uint8_t {
+
+/// Produces a set of all protocols that have an inverse, i.e., for every
+/// known protocol KP in the set, ~KP exists.
+KnownProtocolSet getInvertableProtocols();
```
Typo: invertible not invertable. But I think this isn't a great name. Perhaps it should also be a global constant instead of a function?

slavapestov replied:
What about calling these "removable protocols"?

kavon replied:
I'm going with the pitch's terminology (which I also invented myself) for now to get this prototyped. Previous iterations of this used the term "Suppressible". Given that there's also going to be the notion of subtraction (i.e., `Type ~ Copyable`) to remove defaults coming from `Type`, I think using type inverse/subtraction terminology is at least aligned.
